lib/gssapi/spnego/context_storage.c

```diff
@@ -0,0 +1,584 @@
+        ret = KRB5_BAD_MSIZE;
+        goto fail;
+    }
+
+    if (snc_flags & SNC_MECH_CONTEXT) {
+        CHECK(ret, ret_gss_buffer(sp, &buf));
+        CHECK(major, gss_import_sec_context(&minor, &buf,
+                                            &mech->mech_context));
+        gss_release_buffer(&minor, &buf);
+    }
+
+    if (snc_flags & SNC_METADATA)
+        CHECK(ret, ret_gss_buffer(sp, &mech->metadata));
+
+    *mechp = mech;
+
+fail:
+    if (ret == 0 && GSS_ERROR(major))
+        ret = minor ? minor : KRB5_BAD_MSIZE;
+    if (ret)
+        _gss_negoex_release_auth_mech(context, mech);
+    gss_release_buffer(&minor, &buf);
+
+    return ret;
+}
+
+static krb5_error_code
+store_negoex_auth_mech(krb5_storage *sp, struct negoex_auth_mech *mech)
+{
+    krb5_error_code ret;
+    OM_uint32 major = GSS_S_COMPLETE, minor;
+    gss_buffer_desc buf = GSS_C_EMPTY_BUFFER;
+    uint8_t negoex_flags = 0, snc_flags = 0;
+
+    negoex_flags = 0;
+    if (mech->complete)
+        negoex_flags |= (1 << 0);
+    if (mech->sent_checksum)
+        negoex_flags |= (1 << 1);
+    if (mech->verified_checksum)
+        negoex_flags |= (1 << 2);
+
+    if (mech->oid)
+        snc_flags |= SNC_OID;
+    if (mech->mech_context)
+        snc_flags |= SNC_MECH_CONTEXT;
+    if (mech->metadata.length)
+        snc_flags |= SNC_METADATA;
+
+    CHECK(ret, krb5_store_uint8(sp, snc_flags));
+    CHECK(ret, krb5_store_uint8(sp, negoex_flags));
+
+    if (snc_flags & SNC_OID)
+        CHECK(ret, store_gss_oid(sp, mech->oid));
+
+    CHECK(ret, krb5_store_bytes(sp, mech->scheme, GUID_LENGTH));
+
+    if (snc_flags & SNC_MECH_CONTEXT) {
+        CHECK(major, gss_export_sec_context(&minor, &mech->mech_context,
+                                            &buf));
+        CHECK(ret, store_gss_buffer(sp, &buf));
+    }
+
+    if (snc_flags & SNC_METADATA)
+        CHECK(ret, store_gss_buffer(sp, &mech->metadata));
+
+fail:
+    if (ret == 0 && GSS_ERROR(major))
+        ret = minor ? minor : KRB5_BAD_MSIZE;
+    gss_release_buffer(&minor, &buf);
+
+    return ret;
+}
+
+static krb5_error_code
+ret_gss_oid(krb5_storage *sp, gss_OID *oidp)
+{
+    krb5_data data;
+    krb5_error_code ret;
+    gss_OID_desc oid;
+    OM_uint32 major, minor;
+
+    *oidp = GSS_C_NO_OID;
+
+    ret = krb5_ret_data(sp, &data);
+    if (ret)
+        return ret;
+
+    if (data.length) {
+        oid.length = data.length;
+        oid.elements = data.data;
+
+        major = _gss_intern_oid(&minor, &oid, oidp);
+    } else
+        major = GSS_S_COMPLETE;
+
+    krb5_data_free(&data);
+
+    return GSS_ERROR(major) ? ENOMEM : 0;
+}
+
+static krb5_error_code
+store_gss_oid(krb5_storage *sp, gss_OID oid)
+{
+    krb5_data data;
+
+    krb5_data_zero(&data);
+
+    if (oid) {
+        data.length = oid->length;
+        data.data = oid->elements;
+    }
+
+    return krb5_store_data(sp, data);
+}
+
+static krb5_error_code
+ret_gss_buffer(krb5_storage *sp, gss_buffer_t buffer)
```

Comment:
I wonder if `{ret,store}_gss_{oid,buffer}()` might be of more general use and might want to have a life where they can be shared.

Reply:
`_gss_mg_{ret,store}_gss_{oid,buffer}()` makes sense. Question is whether we should introduce those now or wait until we have a use?